### ISSUE TITLE
Discussion: Remove Memory Browser

### ISSUE THREAD

**Issue by colin-grant-work** [eclipse-cdt/cdt-gdb-vscode#110]
The [VSCode Memory Inspector](https://github.com/eclipse-cdt-cloud/vscode-memory-inspector) is a sister project in CDT Cloud that derives ultimately from an enhancement of the Memory Browser in this repository. Would it make sense to remove the Memory Browser code from this repository and direct users to the Memory Inspector in some way?

@thegecko as a Memory Inspector contributor. @asimgunes as a consumer of the exports here.

**Comment by thegecko** [comment]
I support this proposal and look forward to more features in memory inspector :)

**Comment by jonahgraham** [comment]
_At the [CDT Cloud](https://github.com/eclipse-cdt-cloud/cdt-cloud/wiki/CDT-Cloud-Meetings) call today I was asked to provide additional details. cc: @asimgunes_

I don't recommend that we invest further in the integrated memory browser and invest future work only on the [memory inspector](https://github.com/eclipse-cdt-cloud/vscode-memory-inspector). What prevents removing this in the short term is the bit of code that communicates with the amalgamator to send an additional field to the memory read request that passes which child to use:

![image](https://github.com/eclipse-cdt-cloud/cdt-gdb-vscode/assets/679236/96707631-85dc-4a12-91e6-96af5c1b2b48)

The UI only appears if connected to amalgamator:

https://github.com/eclipse-cdt-cloud/cdt-gdb-vscode/blob/36449f8ba2ce2a2fe47fa5b57adb3e25c3940269/src/memory/client/MemoryBrowser.tsx#L99

But IMHO the correct solution is to have the context (selected frame in the _call stack_) available to other views in the extension. As mentioned on the call today there is some progress on that from VSCode community (references needed)

**Comment by colin-grant-work** [comment]
@jonahgraham, it looks like this plugin has handling for cases where [`debugType == amalgamator`](https://github.com/eclipse-cdt-cloud/cdt-gdb-vscode/blob/36449f8ba2ce2a2fe47fa5b57adb3e25c3940269/package.json#L710-L721), but it doesn't actually contribute [the debug type `amalgamator`](https://github.com/eclipse-cdt-cloud/cdt-gdb-vscode/blob/36449f8ba2ce2a2fe47fa5b57adb3e25c3940269/package.json#L59).  Is that handled by other plugins, or who exposes that debug type to the IDE?

**Comment by jonahgraham** [comment]
>  I wouldn't expect that to get read as a plugin manifest by VSCode.

It may very well not work. That code was supposed to be "us" (Renesas) upstreaming our internal changes, but we still use the internal version so that code in this repo may not actually work.

**Comment by colin-grant-work** [comment]
I think I just misunderstood the mechanism: it looks like the code in this repo depends on the equivalent of an NPM package version of the Amalgamator code, but the Amalgamator is also a plugin, so it does get read as a plugin. Thanks!

**Comment by WyoTwT** [comment]
Hi all. I've been working on some changes to the [VSCode Memory Inspector ](https://github.com/eclipse-cdt-cloud/vscode-memory-inspector)to replace the Memory Browser in VSCode for the Renesas application.

One thing I noticed is that the data return format from the `cdt-gdb-adapter's`  [memoryRequest() in GDBDebugSession.ts](https://github.com/eclipse-cdt-cloud/cdt-gdb-adapter/blob/cd740cd0a72f1be57363ccfe07e21a63a6970dbe/src/GDBDebugSession.ts#L1463) is only slightly different from the `cdt-gdb-adapter's` [ readMemoryRequest() in GDBDebugSession.ts](https://github.com/eclipse-cdt-cloud/cdt-gdb-adapter/blob/cd740cd0a72f1be57363ccfe07e21a63a6970dbe/src/GDBDebugSession.ts#L1625) .  Although the input arguments are quite compatible, the biggest difference is in the return values (`memoryRequest()` returns hex while `readMemoryRequest()` returns base64).

The `cdt-amalgamator` is currently using the `memoryRequest()` for the VSCode's MemoryBrowser [[link](https://github.com/eclipse-cdt-cloud/cdt-gdb-vscode/blob/36449f8ba2ce2a2fe47fa5b57adb3e25c3940269/src/memory/server/MemoryServer.ts#L113)] while Memory Inspector is currently using `readMemoryRequest()` [[link](https://github.com/eclipse-cdt-cloud/vscode-memory-inspector/blob/909bf1529917801857da66965b300ca36ef3e18f/src/plugin/memory-provider.ts#L147)]. 

I can either 
1) Add the `readMemoryRequest()` as a new custom request to `cdt-amalgamator` for consistency with the existing Memory Inspector calls or 
2) Use the existing `memoryRequest()` and convert the return data to base64 in the Memory Inspector's amalgamator abstraction layer (AdapterCapabilities for Memory Inspector) before returning to the Memory Inspector's memory view. This is a little messier but can be done.

Option1 would be my preference.  @colin-grant-work asked me to run it past @asimgunes and @jonahgraham since they're familiar with the Renesas requirements (but I also welcome comments from the rest of this list).

**Comment by thegecko** [comment]
A step towards adding this to the memory inspector in a generic way: https://github.com/eclipse-cdt-cloud/vscode-memory-inspector/pull/152

**Comment by colin-grant-work** [comment]
> A step towards adding this to the memory inspector in a generic way: https://github.com/eclipse-cdt-cloud/vscode-memory-inspector/pull/152

@thegecko, I think [your PR](https://github.com/eclipse-cdt-cloud/vscode-memory-inspector/pull/152) is good and useful, but I don't think it actually addresses either this requirement or the goals of [Thor's PR](https://github.com/eclipse-cdt-cloud/vscode-memory-inspector/pull/133) directly, as the point in both cases is to offer sub-session selection. In the case of the amalgamator, there are two GDB sessions, but they're presented to VSCode as a single session with multiple 'threads', so adding a selector at the session level doesn't actually reach the level desired, and similarly, in Thor's PR, the 'context' is the thread level (matching the amalgamators but with other uses in our context downstream), so below VSCode's notion of session, as well.

**Comment by thegecko** [comment]
> @thegecko, I think https://github.com/eclipse-cdt-cloud/vscode-memory-inspector/pull/152 is good and useful, but I don't think it actually addresses either this requirement or the goals of https://github.com/eclipse-cdt-cloud/vscode-memory-inspector/pull/133 directly, as the point in both cases is to offer sub-session selection. In the case of the amalgamator, there are two GDB sessions, but they're presented to VSCode as a single session with multiple 'threads', so adding a selector at the session level doesn't actually reach the level desired, and similarly, in Thor's PR, the 'context' is the thread level (matching the amalgamator's desired selection level, but with other uses in our context downstream, as well), also below VSCode's notion of session.

Right, but these solutions only use `threads` as a way to hook into the VS Code UI effectively. My proposal is to extend the work to make this pluggable in order to display and select any context as it makes sense for individual debuggers.

**Comment by jonahgraham** [comment]
When this is made pluggable there needs to be a way of the extender intercepting messages to modify them. For the amalgamator case it means adding an addition field `child` - but it could require some other types of changes to the request to pass the selection in the drop-down to the target adapter.

Note that because `child` is used today by the amalgamator doesn't mean that is the best solution. @trongle0504 may have more thoughts on this, but it could be adding the threadId (or frameId) to the read memory request is better and more generic. Using frameid would make ReadMemory (and related) more similar to Evaluate Request so that the correct context can be passed from the UI to the adapter.

**Comment by jreineckearm** [comment]
Appreciate we still need to finish sorting out the amalgamator use case in Memory Inspector.
Just wondering if we should probably start anyway with phasing the built-in memory window out.

In a first iteration we can add a launch config or VS Code setting that disables the memory window and related commands (if any exist). Preference would be the launch config, but I haven't had a look yet if feasible.

That way, debug solutions that have configs/settings managed by other extensions (like in our solution), and that are happy to only use the Memory Inspector, can start with turning the built-in window off. This may reduce confusion for users who currently see multiple memory window flavors. And don't know the history of the built-in window and the Memory Inspector.

Maybe this also provokes more real user feedback for this discussion. Those who'd miss it can turn it back on until we get serious and fully remove it.

Then, when amalgamator got sorted, we can fully remove it and the setting during one of the next major version bumps.
